lib/phoenix/code_reloader.ex

```diff
@@ -49,17 +51,201 @@ defmodule Phoenix.CodeReloader do
   API used by Plug to invoke the code reloader on every request.
   """
   def call(conn, opts) do
-    case opts[:reloader].(conn.private.phoenix_endpoint) do
+    task = Task.async(fn -> recv_loop(conn) end)
+
+    Proxy.forward_to(task.pid)
+
+    res = opts[:reloader].(conn.private.phoenix_endpoint)
+
+    {:ok, {conn, output}} = Task.shutdown(task)
+
+    case res do
       :ok ->
-        conn
-      {:error, output} ->
+        if feedback_started?(conn) do
+          Plug.Conn.chunk(conn, "Done!<br />\n")
+          Plug.Conn.chunk(conn, ~s(<meta http-equiv="refresh" content="0">\n))
+          halt(conn)
+        else
+          conn
+        end
+      :error ->
+        if !feedback_started?(conn) do
+          conn
+          |> put_resp_content_type("text/html")
+          |> send_resp(500, template(output))
+          |> halt()
+        else
+          halt(conn)
+        end
+    end
+  end
+
+  defp recv_loop(conn) do
+    receive do
+      {:done, output} ->
+        {conn, output}
+      {:chars, channel, chars} ->
+        conn = compile_callback(conn, channel, chars)
+        recv_loop(conn)
+    end
+  end
+
+  def compile_callback(conn, channel, chars) do
+    if send_feedback?(conn) do
+      conn = start_progress_output(conn)
+      html = ~s(<span class="#{channel}">) <> Phoenix.CodeReloader.Colors.to_html(chars) <> "</span>"
+      {:ok, conn} = Plug.Conn.chunk(conn, html)
+      conn
+    else
+      conn
+    end
+  end
+
+  defp send_feedback?(conn),
+    do: accepts_html?(conn) and not is_ajax?(conn)
+
+  defp feedback_started?(conn),
+    do: conn.state == :chunked
+
+  defp accepts_html?(conn) do
+    conn
+    |> get_req_header("accept")
+    |> Enum.any?(&String.contains?(&1, "html"))
```

Comment:
I believe `contains` is better here, unless we want to start properly reading the accept field.

E.g. here is the `accept` value I'm testing against:
```"text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8"```

We could parse this properly, but since we only care if HTML is anywhere in the list (or even XHTML, in this case), I think the `String.contains` is the best option.